keycloak/sms-provider/src/main/java/org/sunbird/keycloak/resetcredential/chooseuser/ResetCredentialChooseUserAuthenticator.java

```diff
@@ -0,0 +1,137 @@
+import org.keycloak.models.UserModel;
+import org.keycloak.services.messages.Messages;
+import org.sunbird.keycloak.resetcredential.sms.KeycloakSmsAuthenticator;
+import org.sunbird.utils.SunbirdAuthUtil;
+
+
+/**
+ * 
+ * @author Amit Kumar
+ *
+ *         This class will override the choose user action for reset credential flow under
+ *         Authentication. Here we are overriding action method for getting user details by
+ *         attributes (i.e phone)
+ */
+public class ResetCredentialChooseUserAuthenticator implements Authenticator {
+
+
+  private static Logger logger = Logger.getLogger(KeycloakSmsAuthenticator.class);
+  public static final String PROVIDER_ID = "reset-credentials-choose-user";
+
+
+  @Override
+  public void authenticate(AuthenticationFlowContext context) {
+
+    String existingUserId =
+        context.getAuthenticationSession().getAuthNote(AbstractIdpAuthenticator.EXISTING_USER_INFO);
+    if (existingUserId != null) {
+      UserModel existingUser = AbstractIdpAuthenticator.getExistingUser(context.getSession(),
+          context.getRealm(), context.getAuthenticationSession());
+
+      logger.debugf(
+          "Forget-password triggered when reauthenticating user after first broker login. Skipping reset-credential-choose-user screen and using user '%s' ",
+          existingUser.getUsername());
+      context.setUser(existingUser);
+      context.success();
+      return;
+    }
+
+    String actionTokenUserId =
+        context.getAuthenticationSession().getAuthNote(DefaultActionTokenKey.ACTION_TOKEN_USER_ID);
+    if (actionTokenUserId != null) {
+      UserModel existingUser =
+          context.getSession().users().getUserById(actionTokenUserId, context.getRealm());
+
+      // Action token logics handles checks for user ID validity and user being enabled
+
+      logger.debugf(
+          "Forget-password triggered when reauthenticating user after authentication via action token. Skipping reset-credential-choose-user screen and using user '%s' ",
+          existingUser.getUsername());
+      context.setUser(existingUser);
+      context.success();
+      return;
+    }
+
+    Response challenge = context.form().createPasswordReset();
+    context.challenge(challenge);
+
+  }
+
+  @Override
+  public void action(AuthenticationFlowContext context) {
+    EventBuilder event = context.getEvent();
+    MultivaluedMap<String, String> formData = context.getHttpRequest().getDecodedFormParameters();
+    String username = formData.getFirst("username");
+    if (username == null || username.isEmpty()) {
+      event.error(Errors.USERNAME_MISSING);
+      Response challenge = context.form().setError(Messages.MISSING_USERNAME).createPasswordReset();
+      context.failureChallenge(AuthenticationFlowError.INVALID_USER, challenge);
+      return;
+    }
+    List<UserModel> userModels = SunbirdAuthUtil.getUser(context, username);
+    UserModel user = null;
+    if (null != userModels && userModels.size() > 1) {
```

Comment:
Please see notes in AbstractPhoneFormAuthenticator on handling of duplicate users
[ ] Use `ModelDuplicateException` which indicates which field was used resulting in duplicates
[ ] Externalise strings so that they can be localised
[ ] Use `setDuplicateUserChallenge`